tokio/src/task/join_map.rs

```diff
@@ -0,0 +1,566 @@
+///
+/// When the `JoinMap` is dropped, all tasks in the `JoinMap` are immediately aborted.
+///
+/// **Note**: This is an [unstable API][unstable]. The public API of this type
+/// may break in 1.x releases. See [the documentation on unstable
+/// features][unstable] for details.
+///
+/// [`JoinSet`]: crate::task::JoinSet
+/// [unstable]: crate#unstable-features
+/// [abort]: fn@Self::abort
+/// [abort_matching]: fn@Self::abort_matching
+/// [contains]: fn@Self::contains_task
+#[cfg_attr(docsrs, doc(cfg(all(feature = "rt", tokio_unstable))))]
+pub struct JoinMap<K, V, S = RandomState> {
+    /// A hash map of keys to `AbortHandle`s, used to cancel tasks by key.
+    ///
+    /// Note: this currently stores an `Option<AbortHandle>`, in order to
+    /// implement the `abort_matching` method without adding a temporary
+    /// allocation. Because we need to take the `AbortHandle`s by value to abort
+    /// them, we have to take them out of the map somehow. We cannot do that
+    /// while iterating without also removing the tasks that are _not_ being
+    /// aborted, which would then require that we store those K-V pairs someplace
+    /// and then put them back afterwards (requiring an allocation). This seems
+    /// not great.
+    ///
+    /// Instead, we store the `AbortHandle`s in an `Option`, so that we can
+    /// `take` them in `abort_matching`. When `abort_matching` is called, the
+    /// aborted tasks will still be removed from the map in `poll_join_one`, so
+    /// it's fine to leave a `(Key, None)` pair in the map for a little bit.
+    /// Everywhere else, the `Option` is not `take`n, since the task is just
+    /// removed from the map entirely in order to abort it.
+    ///
+    /// When the `HashMap::drain_filter` API is stabilized in `std`
+    /// (<https://github.com/rust-lang/rust/issues/59618>), we can just
+    /// use that, instead, and remove the `Option`s here.
+    aborts: HashMap<K, Option<AbortHandle>, S>,
+    joins: IdleNotifiedSet<(K, JoinHandle<V>)>,
+}
+
+impl<K, V> JoinMap<K, V> {
+    /// Create a new empty `JoinMap`.
+    ///
+    /// The `JoinMap` is initially created with a capacity of 0, so it will not
+    /// allocate until a task is first spawned on it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio_util::task::JoinMap;
+    /// let mut map: JoinMap<&str, i32> = JoinMap::new();
+    /// ```
+    pub fn new() -> Self {
+        Self::with_hasher(RandomState::new())
+    }
+
+    /// Creates an empty `JoinMap` with the specified capacity.
+    ///
+    /// The `JoinMap` will be able to hold at least `capacity` tasks without
+    /// reallocating. If `capacity` is 0, the `JoinMap` will not allocate.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio_util::task::JoinMap;
+    /// let mut map: JoinMap<&str, i32> = JoinMap::with_capacity(10);
+    /// ```
+    #[inline]
+    #[must_use]
+    pub fn with_capacity(capacity: usize) -> Self {
```

Comment:
because i forgot :) i'll add it.

i noticed that std marks all collection constructors with `#[must_use]` and thought we should consider doing that as well.